core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1525,15 +1540,12 @@ private FileBlockInfo generateFileBlockInfo(LockedInodePath inodePath, BlockInfo
     if (fileBlockInfo.getBlockInfo().getLocations().isEmpty() && file.isPersisted()) {
       // No alluxio locations, but there is a checkpoint in the under storage system. Add the
       // locations from the under storage system.
-      MountTable.Resolution resolution = mMountTable.resolve(inodePath.getUri());
-      String ufsUri = resolution.getUri().toString();
-      UnderFileSystem ufs = resolution.getUfs();
-      List<String> locs;
-      try {
-        locs = ufs.getFileLocations(ufsUri,
+      long blockId = fileBlockInfo.getBlockInfo().getBlockId();
+      List<String> locs = mUfsBlockLocationCache.get(blockId);
+      if (locs == null) {
+        mUfsBlockLocationCache.process(blockId, inodePath.getUri(),
             FileLocationOptions.defaults().setOffset(fileBlockInfo.getOffset()));
-      } catch (IOException e) {
-        return fileBlockInfo;
+        locs = mUfsBlockLocationCache.get(blockId);
```

Comment:
is this guaranteed to be not-null? Instead, maybe `process` should return the results?

Reply:
renamed to `get`